torch/csrc/distributed/rpc/request_callback_no_python.cpp

```diff
@@ -583,25 +502,48 @@ void RequestCallbackNoPython::processRpc(
             wrappedMsgType,
             messageId,
             wrappedRpcResponseFuture);
-      }
-      wrappedRpcResponseFuture->addCallback([wrappedRpcResponseFuture,
+
+        auto tid = std::this_thread::get_id();
+        wrappedRpcResponseFuture->addCallback(
+            at::wrapPropagateTLSState<void>([wrappedRpcResponseFuture,
                                              responseFuture,
-                                             profiledEvents =
-                                                 std::move(profiledEvents),
-                                             profilingKeyId] {
-        if (wrappedRpcResponseFuture->hasError()) {
-          // Propagate error
-          responseFuture->setError(wrappedRpcResponseFuture->error()->what());
-        } else {
-          auto rpcWithProfilingResp = std::make_unique<RpcWithProfilingResp>(
-              MessageType::RUN_WITH_PROFILING_RESP,
-              std::move(*wrappedRpcResponseFuture).moveValue(),
-              profiledEvents,
-              profilingKeyId);
-          responseFuture->markCompleted(
-              std::move(*rpcWithProfilingResp).toMessage());
-        }
-      });
+                                             profilingKeyId,
+                                             profilingConfig,
+                                             tid] {
+              std::vector<torch::autograd::profiler::Event> profiledEvents;
+              // Defer consolidation of profiler events until async work has
+              // completed (such as async UDF)
+
+              TORCH_INTERNAL_ASSERT(
+                  torch::autograd::profiler::profilerEnabled(),
+                  "Expected profiler to be enabled!");
+
+              // Only clean up TLS states of profiler if we are disabling on
```

Comment:
qq, would I be correct if I assume the S in TLS is for Storage instead of States?

Reply:
I've been referring to it as "thread local state", but just realized that this will expand to "thread local state state", which is redundant. I guess it should be "TL states". Will clean these up in a future PR if you don't mind.

Reply:
Sure, not a problem at all. Just please also refactor var names as well.